### PR TITLE
Feature/boas 482 create and display sub folders

### PR DESCRIPTION
**Pull request by DavidJWhiteDev**
## Describe your changes

API - add functionality to create the full folder structure for Applications type cases, when the case moves from Draft to Pre-Application.  Currently 83 folders in total.
Web - changes to correctly display the subfolders in an accordion, and allow traversal through the folders

## BOAS-482 Create and Display sub folders on Applications Case
https://pins-ds.atlassian.net/jira/software/projects/BOAS/boards/172?selectedIssue=BOAS-482

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [x] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [x] I have provided a description of the ticket
- [x] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: apps/web/src/server/views/applications/case-documentation/documentation-folder.njk [hunk 1]
```diff
@@ -12,17 +14,28 @@
...
-	{% endfor %}
+		{% set folderRows = [] %}
+		{% for subFolder in subFolders %}
+			{% set folderRows = folderRows | concat ([[
+				{ html: '<a class="govuk-link govuk-body-m" href="' + 'document-category' |url({caseId: caseId, documentationCategory: subFolder}) +'">' + subFolder.displayNameEn + '</a>' }
+			]])
+			%}
+		{% endfor %}
+
+		{{ govukAccordion({
```
Comment (francescobrenci-AR) [inline apps/web/src/server/views/applications/case-documentation/documentation-folder.njk:29]: do you need an Accordion here? I can see in the CSS that the controls are then hidden

File: apps/web/src/server/views/applications/case-documentation/documentation-folder.njk [hunk 2]
```diff
@@ -37,13 +50,18 @@
...
+	<h2 class="govuk-heading-m">Documents</h2>
 	<!-- TODO: loop to display files in this folder -->
-	{% for documentationFile in documentationFiles %}
-		<div class="govuk-grid-row govuk-grid-column-full">
-			<a class="govuk-link govuk-body-s govuk-!-font-weight-bold" href={{ documentationFile.url }}>
-				{{ documentationFile.fileName }}
-			</a>
-		</div>
-	{% endfor %}
+	{% if documentationFiles |length %}
```
Comment (francescobrenci-AR) [inline apps/web/src/server/views/applications/case-documentation/documentation-folder.njk:55]: It makes sense to use a list here:

{% if documentationFiles |length %}
<ul>
  {% for documentationFile in documentationFiles %}
   <li class="govuk-grid-row govuk-grid-column-full">
	<a class="govuk-link govuk-body-s govuk-!-font-weight-bold" href={{ documentationFile.url }}>
		{{ documentationFile.fileName }}
	</a>
    </li>
    {% endfor %}
</ul>
{% else %}

**Review by francescobrenci-AR — CHANGES_REQUESTED** [review]

File: apps/api/src/server/repositories/folder.repository.js [hunk 3]
```diff
@@ -69,39 +71,274 @@ export const getFolderPath = async (caseId, currentFolderId) => {
...
+
+	return foldersCreated;
 };
 
+/**
+ * The default template folder structure for a new NI Applications type case
+ *
+ * @type {FolderTemplate[]} defaultCaseFolders
+ */
 export const defaultCaseFolders = [
```
Comment (mayya-bond-kc) [inline apps/api/src/server/repositories/folder.repository.js:124]: I don't believe we need to export this anymore

File: apps/api/src/server/repositories/folder.repository.js [hunk 4]
```diff
@@ -69,39 +71,274 @@ export const getFolderPath = async (caseId, currentFolderId) => {
...
 
 /**
+ * Creates the top level folders on a case using the folder template
+ * and recursively creates all sub folders.
+ * Returns an array of promises
  *
- * @param {number} folderId
- * @returns {Promise<import('@pins/api').Schema.Folder | null>}
+ * @param {number} caseId
+ * @returns {Promise<import('@pins/api').Schema.BatchPayload>[]}
```
Comment (mayya-bond-kc) [inline apps/api/src/server/repositories/folder.repository.js:98]: I did not realise that the response of a `folder.create` function is a `Promise<BatchPayload>`? I thought it was `Promise<Folder>`?

File: apps/api/src/server/applications/application/__tests__/start-case.test.js [hunk 5]
```diff
@@ -127,21 +127,28 @@ test('starts application if all needed information is present', async (t) => {
...
+								{ displayNameEn: 'Travel', displayOrder: 100, caseId: 1 },
+								{ displayNameEn: 'Welsh', displayOrder: 200, caseId: 1 }
+							]
+						}
+					},
+					{ displayNameEn: 'Mail merge spreadsheet', displayOrder: 200, caseId: 1 },
+					{ displayNameEn: 'Fees', displayOrder: 300, caseId: 1 }
+				]
+			}
+		}
```
Comment (mayya-bond-kc) [inline apps/api/src/server/applications/application/__tests__/start-case.test.js:151]: Is the plan only to test that a single top-most level folder gets created but not the others?

File: apps/api/src/server/repositories/folder.repository.js (hunk 4 above)
Comment (DavidJWhiteDev) [reply apps/api/src/server/repositories/folder.repository.js:98]: I don't know Mayya.  It was returning a BatchPayLoad before, so I did not change it, I just made it an array [].  The compare above looks a bit confusing, I think it is showing the params for a separate fn.  I looked back through code before this ticket and it was BatchPayLoad.

Comment (mayya-bond-kc) [reply apps/api/src/server/repositories/folder.repository.js:98]: I think `BatchPayload` was accurate as of using `createMany`, but not any longer as we are using `create`

Comment (DavidJWhiteDev) [reply apps/api/src/server/repositories/folder.repository.js:98]: ok, I will change to Promise<Folder>[] and re-test

**Review by mayya-bond-kc — APPROVED** [review]

**Review by francescobrenci-AR — APPROVED** [review]

**Pull request merged**